File: packages/python/src/xyp/_tls.py
"""TLS trust for xyp.gov.mn.

XYP's certificate is issued by the Mongolian national CA, which no operating
system or Python trust store ships. Instead of turning verification off (what
most integrations do), we trust exactly the national root + issuing CA and
nothing else. See docs/tls.md in the repository.
"""

from __future__ import annotations

import os
import ssl
from importlib import resources

Verify = bool | str | os.PathLike[str] | ssl.SSLContext

_BUNDLED_CERTIFICATES = ("MNRCA-2021.pem", "MNICA-2022.pem")


def bundled_ca_pem() -> str:
    certs = resources.files("xyp") / "certs"
    return "".join((certs / name).read_text(encoding="ascii") for name in _BUNDLED_CERTIFICATES)


def build_verify(verify: Verify) -> ssl.SSLContext | bool:
    """Translate the client's `verify` option into what httpx expects.

    True (default) -> bundled Mongolian national CAs only
    path           -> your own CA bundle (PEM file)
    SSLContext     -> used as is
    False          -> no verification; only for proxies/tests, never the default
    """
    if isinstance(verify, ssl.SSLContext):
        return verify
    if verify is False:
        return False
    if verify is True:
        context = ssl.create_default_context(cadata=bundled_ca_pem())
        return context
    return ssl.create_default_context(cafile=os.fspath(verify))

File: packages/go/xyp/tls.go
package xyp

import (
	"crypto/tls"
	"crypto/x509"
	_ "embed"
	"net/http"
)

// XYP's certificate is issued by the Mongolian national PKI, which no operating
// system or Go trust store ships. Instead of turning verification off (what most
// integrations do, including the official sample), the SDK trusts exactly the
// national root and issuing CA, and only for its own requests. The two files are
// byte-for-byte the ones published at https://esign.gov.mn/MNRCA.zip and
// https://esign.gov.mn/MNICA.zip. See docs/tls.md in the repository.
var (
	//go:embed certs/MNRCA-2021.pem
	nationalRootCA string
	//go:embed certs/MNICA-2022.pem
	nationalIssuingCA string
)

// BundledCAs returns a fresh pool holding MNRCA-2021 and MNICA-2022 and nothing
// else. Add your own certificates to it and pass it as Options.RootCAs when you
// reach XYP through a proxy that re-terminates TLS.
func BundledCAs() *x509.CertPool {
	pool := x509.NewCertPool()
	pool.AppendCertsFromPEM([]byte(nationalRootCA))
	pool.AppendCertsFromPEM([]byte(nationalIssuingCA))
	return pool
}

// newTransport clones the default transport so connection pooling, proxy support
// and HTTP/2 behave as they do everywhere else, and only replaces the trust.
func newTransport(roots *x509.CertPool, insecure bool) *http.Transport {
	transport, isDefault := http.DefaultTransport.(*http.Transport)
	if !isDefault {
		transport = &http.Transport{}
	}
	cloned := transport.Clone()
	cloned.TLSClientConfig = &tls.Config{
		MinVersion:         tls.VersionTLS12,
		RootCAs:            roots,
		InsecureSkipVerify: insecure, //nolint:gosec // opt-in only, via Options.InsecureSkipVerify
	}
	return cloned
}
